### PR TITLE
feat: BIP341/BIP342 SegWit V1 SigHash + extension

### PR DESCRIPTION
**Pull request by brandonblack**

### CODE REVIEW

File: ts_src/transaction.ts [hunk 1]
```diff
@@ -336,6 +340,156 @@ export class Transaction {
...
+  ): Buffer {
+    // https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message
+    typeforce(
+      types.tuple(
+        types.UInt32,
+        typeforce.arrayOf(types.Buffer),
+        typeforce.arrayOf(types.Satoshi),
+        types.UInt32,
+      ),
+      arguments,
```
Comment (sangaman) [inline ts_src/transaction.ts:359]: This is an odd pattern to me but I guess it's how it's used elsewhere, like I don't know where `arguments` comes from. I like the way it's done in `payments` better with the named parameters, but I guess it's better to keep as-is for consistency with the rest of the file.

File: ts_src/transaction.ts [hunk 2]
```diff
@@ -336,6 +340,156 @@ export class Transaction {
...
+    const spendType = (leafHash ? 2 : 0) + (annex ? 1 : 0);
+
+    // Length calculation from:
+    // https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-14
+    const sigMsgSize =
+      174 -
+      (isAnyoneCanPay ? 49 : 0) -
+      (isNone ? 32 : 0) +
+      (annex ? 32 : 0) +
+      (leafHash ? 37 : 0);
```
Comment (sangaman) [inline ts_src/transaction.ts:443]: I didn't see leaf hash in the cited note, is this the part you mentioned that doesn't get covered until BIP342? I see the 37 bytes written a few lines below.

**Review by sangaman — APPROVED** [review]

File: ts_src/transaction.ts (hunk 2 above)
Comment (brandonblack) [reply ts_src/transaction.ts:443]: Yeah, exactly, it's not mentioned in the size calculation. I'll improve the comment.

File: ts_src/transaction.ts (hunk 1 above)
Comment (brandonblack) [reply ts_src/transaction.ts:359]: Yeah, totally agreed.

**Pull request merged**